### PR TITLE
ds: Support renaming of a replication plugin in 389-ds

### PR DESCRIPTION
**Pull request by abbra**
IPA topology plugin depends on the replication plugin but
389-ds cannot handle older alias querying in the plugin
configuration with 'nsslapd-plugin-depends-on-named: ..' attribute

See https://github.com/389ds/389-ds-base/issues/4786 for details

Fixes: https://pagure.io/freeipa/issue/8799

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

### CODE REVIEW

**Comment by abbra** [comment]
I verified in a separe private Azure CI run that 389-ds nightly on Rawhide works just fine with this PR, so we cover both 'old' 389-ds and 'new' 389-ds. The only problem is that on Rawhide we cannot complete replica deployment with CA due to https://bugzilla.redhat.com/show_bug.cgi?id=1965721 in Dogtag, unrelated to this change.

**Comment by stanislavlevin** [comment]
If I understand correctly this is the static configuration, the plugin name is detected either on server/replica installation or the IPA upgrade procedure.

Is the next case considered?
1) old 389
2) IPA + this patch
3) upgrade 389 RPM
4) old plugin name :( until IPA upgrade

**Comment by stanislavlevin** [comment]
Is there no IPA solution for https://github.com/freeipa/freeipa/pull/5739#issuecomment-843208907?

**Comment by stanislavlevin** [comment]
What will be if some of domain servers/replicas have old plugin name and some - new?

**Comment by abbra** [comment]
@stanislavlevin 389-ds does rename own plugins on each startup, so it will handle the case. However, it currently doesn't handle the case when a plugin already renamed but the dependencies aren't renamed -- see my latest comment in the 389-ds issue. This is an issue, I believe, that should be solved on 389-ds side.

> Is there no IPA solution for [#5739 (comment)](https://github.com/freeipa/freeipa/pull/5739#issuecomment-843208907)?

This patchset solves this problem. If IPA is installed with either old or new 389-ds, we detect what plugin name is used and adjust ourselves. So this is good.



> What will be if some of domain servers/replicas have old plugin name and some - new?
Since all plugin names are part of a non-replicated space (`cn=config`), we don't need to care about other replicas. Plugin names aren't referenced on the wire.

**Comment by tiran** [comment]
There is a much easier way:

## 2.0.4
```
>>> lib389.config.PLUGIN_REPLICATION
'Multimaster Replication Plugin'
```

## 2.0.5
```
>>> lib389.config.PLUGIN_REPLICATION
'Multisupplier Replication Plugin'
```

**Comment by abbra** [comment]
I do not want to depend on lib389 here. For tests we want to get rid of lib389 dependency as it is forced on a test controller, not just SUT.

**Comment by abbra** [comment]
I updated Fedora 34 bodhi request with a fixed FreeIPA version that includes the patch from this PR and triggered OpenQA run:
https://openqa.fedoraproject.org/tests/overview?build=Update-FEDORA-2021-234a7eff94&distri=fedora&version=34&groupid=2
this test includes an upgrade situation from F33 to F34, so we will see how it survives this update.

File: .wheelconstraints.in [hunk 1]
```diff
@@ -12,4 +12,6 @@ ipatests == @VERSION@
 # keep pylint version in sync with current Fedora release
 # F34 has 2.6
 # https://pagure.io/freeipa/issue/8818 added pylint 2.8 support
+# but somehow 2.8.2 still breaks us
 pylint < 2.9
+astroid < 2.5.7
```
Comment (stanislavlevin) [inline .wheelconstraints.in:17]: no longer required with Pylint 2.8.3.

File: ipaserver/install/replication.py [hunk 2]
```diff
@@ -232,6 +232,23 @@ def get_ds_version(conn):
...
+    # configuration with 'nsslapd-plugin-depends-on-named: ..' attribute
+    #
+    # dirsrv_get_entry: function (dn, attrs) -> str
+    # returns entry dictionary
+    try:
+        entry = dirsrv_get_entry(
+            "cn=Multisupplier Replication Plugin,cn=plugins,cn=config",
+            ["cn"])
+        return entry['cn']
+    except Exception:
```
Comment (stanislavlevin) [inline ipaserver/install/replication.py:248]: can the Exception be more specific?

Comment (abbra) [reply ipaserver/install/replication.py:248]: Unfortunately, no. There are two different `get_entry` implementations (one from IPA, one from lib389) and they throw different exceptions.

File: .wheelconstraints.in (hunk 1 above)
Comment (abbra) [reply .wheelconstraints.in]: I removed the commit.

**Comment by stanislavlevin** [comment]
There is the pattern 'check supplier, if not or exception - use master', can we explicitly check the latter in such cases?

**Comment by abbra** [comment]
I'd prefer not to. Our aim is to eventually retire 'master' / 'slave' / etc terminology and use 'supplier' instead. Thus, going forward, I'd like to reduce 'master' term to absolute minimum.

**Comment by stanislavlevin** [comment]
If some error occured and it is not related to the name of plugin(code error, runtime error), the resulted name of plugin may be wrong.

**Comment by abbra** [comment]
Yes, that's true. However, if that happens, a deployment would be broken anyway.

**Comment by stanislavlevin** [comment]
btw, @tiran 's suggestion doesn't require LDAP connection and the query since `PLUGIN_REPLICATION` is taken from constants.
`ipaserver/install/dsinstance.py` already has dependency on lib389, only `ipaserver/install/replication.py` requires to add one.

**Comment by abbra** [comment]
https://openqa.fedoraproject.org/tests/overview?distri=fedora&version=34&build=Update-FEDORA-2021-234a7eff94&groupid=2 succeeded with the current version of the patchset, including F33 to F34 upgrade.

**Comment by stanislavlevin** [comment]
@abbra, either way it works, no objections. Thank you!

**Comment by abbra** [comment]
master:

* b4b2c10e234315042b54a4556bf264d55ecd07c9 ds: Support renaming of a replication plugin in 389-ds

**Pull request closed without merging**